python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py

```diff
@@ -1860,6 +1863,364 @@ def delete_dynamic_partition(self, partitions_def_name: str, partition_key: str)
+                        PendingStepsTable.c.step_key == step_key,
+                        PendingStepsTable.c.concurrency_key == concurrency_key,
+                    )
+                )
+            ).fetchone()
+            return row and row[0] is not None
+
+    def has_pending_step(self, concurrency_key: str, run_id: str, step_key: str):
+        with self.index_connection() as conn:
+            row = conn.execute(
+                db.select([db.func.count()])
+                .select_from(PendingStepsTable)
+                .where(
+                    db.and_(
+                        PendingStepsTable.c.concurrency_key == concurrency_key,
+                        PendingStepsTable.c.run_id == run_id,
+                        PendingStepsTable.c.step_key == step_key,
+                    )
+                )
+            ).fetchone()
+            return row and cast(int, row[0]) > 0
+
+    def assign_pending_steps(self, concurrency_keys: Sequence[str]):
+        if not concurrency_keys:
+            return
+
+        with self.index_connection() as conn:
+            for key in concurrency_keys:
+                row = conn.execute(
+                    db.select([PendingStepsTable.c.id])
+                    .where(
+                        db.and_(
+                            PendingStepsTable.c.concurrency_key == key,
+                            PendingStepsTable.c.assigned_timestamp == None,  # noqa: E711
+                        )
+                    )
+                    .order_by(
+                        PendingStepsTable.c.priority.desc(),
+                        PendingStepsTable.c.create_timestamp.asc(),
+                    )
+                    .limit(1)
+                ).fetchone()
+                if row:
+                    conn.execute(
+                        PendingStepsTable.update()
+                        .where(PendingStepsTable.c.id == row[0])
+                        .values(assigned_timestamp=db.func.now())
+                    )
+
+    def add_pending_step(
+        self,
+        concurrency_key: str,
+        run_id: str,
+        step_key: str,
+        priority: Optional[int] = None,
+        should_assign: bool = False,
+    ):
+        with self.index_connection() as conn:
+            try:
+                conn.execute(
+                    PendingStepsTable.insert().values(
+                        [
+                            dict(
+                                run_id=run_id,
+                                step_key=step_key,
+                                concurrency_key=concurrency_key,
+                                priority=priority or 0,
+                                assigned_timestamp=db.func.now() if should_assign else None,
+                            )
+                        ]
+                    )
+                )
+            except db_exc.IntegrityError:
+                # do nothing
+                pass
```

Comment:
this is unique contraint guard + no better way to do it across supported DBs?

Reply:
The pattern I've seen is to override with the postgres/mysql variants.  But I think this is the generic way to do it.